crates/astria-sequencer/src/authority/component.rs

```diff
@@ -0,0 +1,79 @@
+use std::sync::Arc;
+
+use anyhow::{
+    Context,
+    Result,
+};
+use proto::native::sequencer::v1alpha1::Address;
+use tendermint::{
+    abci::request::{
+        BeginBlock,
+        EndBlock,
+    },
+    validator,
+};
+use tracing::instrument;
+
+use super::state_ext::{
+    StateReadExt,
+    StateWriteExt,
+    ValidatorSet,
+};
+use crate::component::Component;
+
+#[derive(Default)]
+pub(crate) struct AuthorityComponent;
+
+#[derive(Debug)]
+pub(crate) struct AuthorityComponentAppState {
+    pub(crate) authority_sudo_key: Address,
+    pub(crate) genesis_validators: Vec<validator::Update>,
+}
+
+#[async_trait::async_trait]
+impl Component for AuthorityComponent {
+    type AppState = AuthorityComponentAppState;
+
+    #[instrument(name = "AuthorityComponent:init_chain", skip(state))]
+    async fn init_chain<S: StateWriteExt>(mut state: S, app_state: &Self::AppState) -> Result<()> {
+        // set sudo key and initial validator set
+        state
+            .put_sudo_address(app_state.authority_sudo_key)
+            .context("failed to set sudo key")?;
+        state
+            .put_validator_set(ValidatorSet(app_state.genesis_validators.clone()))
+            .context("failed to set validator set")?;
+        Ok(())
+    }
+
+    #[instrument(name = "AuthorityComponent:begin_block", skip(_state))]
+    async fn begin_block<S: StateWriteExt + 'static>(
+        _state: &mut Arc<S>,
+        _begin_block: &BeginBlock,
+    ) {
+    }
+
+    #[instrument(name = "AuthorityComponent:end_block", skip(state))]
+    async fn end_block<S: StateWriteExt + StateReadExt + 'static>(
+        state: &mut Arc<S>,
+        _end_block: &EndBlock,
+    ) {
+        // update validator set
+        let validator_updates = state
+            .get_validator_updates()
+            .await
+            .expect("failed getting validator updates");
```

Comment:
Should this really crash the entire app?

Reply:
unfortunately, yeah it should. any error during block execution like this will result in the execution ending early, and will cause a resulting different app state which won't match other nodes that did not error, causing a consensus mismatch. there isn't really any way to recover from something like this

Reply:
I guess what I am asking is why this warrants a crash instead of bubbling up the error just taking down tower-abci instead of doing it here with a panic? This would require changing `App::end_block` to return a `Result`. It's actually surprising that some of `App`'s methods are returning errors, while others are outright panicking.

Reply:
@SuperFluffy updated `end_block` to return a `Result`, so the error now bubbles up and is returned to tower-abci